src/masternodeman.cpp

```diff
@@ -502,29 +506,61 @@ CMasternode* CMasternodeMan::GetNextMasternodeInQueueForPayment(int nBlockHeight
         //make sure it has as many confirmations as there are masternodes
         if (mn.GetMasternodeInputAge() < nMnCount) continue;
 
-        vecMasternodeLastPaid.push_back(make_pair(mn.SecondsSincePayment(), mn.vin));
+        if (nBlockHeight < TIER_BLOCK_HEIGHT) {
+            vecMasternodeLastPaid.push_back(make_pair(mn.SecondsSincePayment(), mn.vin));
+        }
+        else {
+            vecMasternodeTiers[mn.tier].push_back(make_pair(mn.SecondsSincePayment(), mn.vin));
+        }
+        nCount++;
     }
 
-    nCount = (int)vecMasternodeLastPaid.size();
-
     //when the network is in the process of upgrading, don't penalize nodes that recently restarted
     if (fFilterSigTime && nCount < nMnCount / 3) return GetNextMasternodeInQueueForPayment(nBlockHeight, false, nCount);
 
-    // Sort them high to low
-    sort(vecMasternodeLastPaid.rbegin(), vecMasternodeLastPaid.rend(), CompareLastPaid());
+    uint256 blockHash = 0;
+    if (!GetBlockHash(blockHash, nBlockHeight - 100)) {
+        LogPrint("masternodeman", "GetNextMasternodeInQueueForPayment ERROR - nHeight %d\n", nBlockHeight - 100);
+    }
+    if (nBlockHeight < TIER_BLOCK_HEIGHT) {
+    //Single tier - 1K
+        pBestMasternode = GetWinningNode(vecMasternodeLastPaid, blockHash);
+    }
+    else {
+    //Apply tier distribution algorithm
+        std::vector<size_t> vecSizes;
+        for (auto i = 0; i < MasternodeTiers::TIER_NONE; i++) {
+            vecSizes.push_back(vecMasternodeTiers[i].size());
+        }
+
+        auto nTier = CalculateWinningTier(vecSizes, blockHash);
+        if (nTier != MasternodeTiers::TIER_NONE) {
+            pBestMasternode = GetWinningNode(vecMasternodeTiers[nTier], blockHash);
+        }
+    }
+    return pBestMasternode;
+}
 
+CMasternode* CMasternodeMan::GetWinningNode(std::vector<pair<int64_t, CTxIn>>& vecMasternodeLastPaid, uint256 blockHash)
+{
+    CMasternode* pBestMasternode = NULL;
+
+    // Sort vector of last paid times high to low
+    if (vecMasternodeLastPaid.size() > 1) {
+        sort(vecMasternodeLastPaid.rbegin(), vecMasternodeLastPaid.rend(), CompareLastPaid());
```

Comment:
what the size of this vector? is it ok that we should sort it all the time?

Reply:
It is the number of masternodes in a tier. It should be sorted every time and it is their legacy logic for node choosing.